test/test-types.mjs

```diff
@@ -0,0 +1,90 @@
+import { test } from 'tap';
+import { execSync } from 'node:child_process';
+import { join } from 'node:path';
+import { fileURLToPath } from 'node:url';
+import { writeFileSync } from 'node:fs';
+
+const __dirname = fileURLToPath(new URL('.', import.meta.url));
+
+// Only run tests in ESM mode (skip when run as CJS from dist/)
+const isESM = !__dirname.includes('/dist/');
+
+test('types: TypeScript compilation with ESM types', { skip: !isESM }, (t) => {
```

Comment:
This ends up making the skip very noisy. Is there a better pattern we could do? Maybe run the ESM tests in the ESM suite and the CJS tests in the CJS suite?